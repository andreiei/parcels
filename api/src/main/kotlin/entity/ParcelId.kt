package entity

import kotlinx.serialization.Serializable

@JvmInline
@Serializable
value class ParcelId(private val value: String) {
    override fun toString(): String = value
}
