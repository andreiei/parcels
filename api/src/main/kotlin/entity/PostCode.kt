package entity

import kotlinx.serialization.Serializable

@JvmInline
@Serializable
value class PostCode(private val value: String) {
    override fun toString(): String = value
}
