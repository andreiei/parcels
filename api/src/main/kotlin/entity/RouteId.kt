package entity

import kotlinx.serialization.Serializable

@JvmInline
@Serializable
value class RouteId(private val value: String) {
    override fun toString(): String = value
}
