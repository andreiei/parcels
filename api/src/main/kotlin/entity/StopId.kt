package entity

import kotlinx.serialization.Serializable

@JvmInline
@Serializable
value class StopId(private val value: String) {
    override fun toString(): String = value
}
