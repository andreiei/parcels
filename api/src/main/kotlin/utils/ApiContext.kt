package utils

import services.ParcelRepository
import services.RoutesService

data class ApiContext(
    val parcelRepository: ParcelRepository = ParcelRepository(),
    val routesService: RoutesService = RoutesService(),
)
