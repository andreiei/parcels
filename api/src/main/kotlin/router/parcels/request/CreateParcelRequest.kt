package router.parcels.request

import entity.ParcelId
import entity.PostCode
import kotlinx.serialization.Serializable

@Serializable
data class CreateParcelRequest(
    val id: ParcelId,
    val weight: Double,
    val postCode: PostCode,
    val address: String,
    val status: ParcelStatus,
) {
    fun validate() {
        require(weight in 0.1..1000.0) {
            "Weight must be between 0.1 and 1000.0 kg"
        }

        require(postCode.toString().matches(Regex("\\d{4}"))) {
            "Post code must be a 4-digit number"
        }

        require(address.length in 1..255) {
            "Address must be between 1 and 255 characters"
        }
    }
}

enum class ParcelStatus {
    NOT_STARTED,
    IN_TRANSIT,
    DELIVERED,
}
