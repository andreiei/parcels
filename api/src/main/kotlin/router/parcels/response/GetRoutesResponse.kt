package router.parcels.response

import entity.ParcelId
import entity.RouteId
import entity.StopId
import kotlinx.serialization.Serializable

@Serializable
data class GetRoutesResponse(
    val routes: List<Route>,
)

@Serializable
data class Route(
    val id: RouteId,
    val stops: List<Stop>,
)

@Serializable
data class Stop(
    val id: StopId,
    val address: String,
    val parcels: List<ParcelId>,
)
