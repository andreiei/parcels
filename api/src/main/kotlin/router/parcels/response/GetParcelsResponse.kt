import entity.ParcelId
import entity.PostCode
import kotlinx.serialization.Serializable
import router.parcels.request.ParcelStatus

@Serializable
data class GetParcelsResponse(
    val parcels: List<ParcelResponse>,
)

@Serializable
data class ParcelResponse(
    val id: ParcelId,
    val status: ParcelStatus,
    val weight: Double,
    val postCode: PostCode,
    val address: String,
)
