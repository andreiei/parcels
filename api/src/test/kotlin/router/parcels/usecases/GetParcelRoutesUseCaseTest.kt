package router.parcels.usecases

import decodeFromString
import entity.ParcelId
import entity.PostCode
import entity.RouteId
import entity.StopId
import io.kotest.core.spec.style.ShouldSpec
import io.mockk.mockk
import router.parcels.request.ParcelStatus
import router.parcels.response.GetRoutesResponse
import router.parcels.usecases.GetRoutesUseCase.Companion.distributeToRoutes
import services.ParcelRepository.Companion.ParcelStorageObject
import shouldMatchSnapshot
import utils.ApiContext

class GetParcelRoutesUseCaseTest : ShouldSpec() {
    init {
        context("distributeToRoutes") {
            val routes: List<RouteId> = listOf(RouteId("R001"), RouteId("R002"), RouteId("R003"))

            should("be handle empty parcels") {
                emptyList<ParcelStorageObject>().distributeToRoutes(routes)
            }
            should("be able to distribute parcels of varying sizes") {
                (0..9)
                    .associateWith { index ->
                        (0..index).map { createParcel(it) }
                            .distributeToRoutes(routes)
                    }
                    .shouldMatchSnapshot(this)
            }
        }
        context(GetRoutesUseCase::invoke.name) {
            val apiContext = ApiContext()
            (0..10)
                .forEach { index ->
                    when (index) {
                        0 -> createParcel(index, "2nd Street")
                        else -> createParcel(index)
                    }
                        .also { apiContext.parcelRepository.addParcel(it) }
                }

            val useCase = GetRoutesUseCase(apiContext)
            val response: GetRoutesResponse = decodeFromString(useCase(mockk()).body!!)
            response
                .routes
                .map { route ->
                    route.copy(stops = route.stops.map { it.copy(id = StopId("REDACTED")) })
                }
                .shouldMatchSnapshot(this)
        }
    }

    private fun createParcel(id: Int, address: String = "123 Main St"): ParcelStorageObject {
        return ParcelStorageObject(
            id = ParcelId("P0$id"),
            weight = 2.3,
            postCode = PostCode("0465"),
            address = address,
            status = ParcelStatus.IN_TRANSIT,
        )
    }
}
