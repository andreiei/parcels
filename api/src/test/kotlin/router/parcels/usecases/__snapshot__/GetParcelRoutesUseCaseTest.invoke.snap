[
    {
        "id": "R001",
        "stops": [
            {
                "id": "REDACTED",
                "address": "2nd Street",
                "parcels": [
                    "P00"
                ]
            },
            {
                "id": "REDACTED",
                "address": "123 Main St",
                "parcels": [
                    "P05",
                    "P010"
                ]
            }
        ]
    },
    {
        "id": "R002",
        "stops": [
            {
                "id": "REDACTED",
                "address": "123 Main St",
                "parcels": [
                    "P01",
                    "P06"
                ]
            }
        ]
    },
    {
        "id": "R003",
        "stops": [
            {
                "id": "REDACTED",
                "address": "123 Main St",
                "parcels": [
                    "P02",
                    "P07"
                ]
            }
        ]
    },
    {
        "id": "R004",
        "stops": [
            {
                "id": "REDACTED",
                "address": "123 Main St",
                "parcels": [
                    "P03",
                    "P08"
                ]
            }
        ]
    },
    {
        "id": "R005",
        "stops": [
            {
                "id": "REDACTED",
                "address": "123 Main St",
                "parcels": [
                    "P04",
                    "P09"
                ]
            }
        ]
    }
]